## Compound Heterozygous Variant Pipeline

### Background
A compound heterozygous variant occurs when a person inherits a variant from one parent within a specific gene and also inherits another variant from the other parent at a different position within the same gene. The effect of compound heterozygotic inheritance results in two recessive alleles that may cause disease. To detect these types of variants it is necessary to differentiate between paternally and maternally derived nucleotides. If sequencing has already taken place, computational algorithms can be used to help determine which nucleotides were inherited from each parent through a process termed “phasing”. 

Phasing requires specific file types which may vary depending on phasing software. Many phasing programs require that input files have been aligned to a specific reference genome, do not contain multiallelic positions, are free of repeat positions, and that each chromosome is phased separately. Figuring out how to prepare files for phasing can be challenging as passing files from program to program may invoke unforeseen incompatibilities. Also, installing specific programs can be challenging because many programs require various dependencies.

### Our Methodology
We have designed our Compound Heterozygous Variant Pipeline to overcome many of the time-consuming challenges that researchers may face when trying to phase patient data. By incorporating our computational pipeline into Docker images and providing reproducible scripts of our analyses, other researchers will be able to examine our methodology in detail and apply it to their data. Encapsulating our code within containers will help control what software versions are used, what system libraries are used, and create a cohesive computational environment.

Currently, this pipeline is designed to be used with data available from the Gabriella Miller Kids First Data Resource Center (GMKFDRC) which is a controlled access repository. Much of the data in the GMKFDRC contains WGS data for the proband and both parents. Our pipeline specifically uses gVCF files that were previously processed using a harmonized pipeline across all disease types. In future updates, we will expand our pipeline to be compatible with data outside the GMKFDRC. Doing so will make our pipeline available to more users and will allow us to provide example data files with example outputs for our users to reference. Please see [ch_pipeline.pdf](https://github.com/dmiller903/ch-pipeline/blob/master/ch_pipeline.pdf) for example code and a description of each step of the pipeline.
